File: ios/XiriGoEcommerceTests/Accessibility/AccessibilityTests.swift
import Foundation
import Testing

/// Static accessibility audit tests that scan source code for common issues.
/// Maps to TEST.md Section 8 — Accessibility Test Layer.
///
/// Runtime accessibility audits (performAccessibilityAudit) require XCUITest.
/// These tests catch missing accessibility modifiers at the code level.
@Suite("Accessibility Tests")
struct AccessibilityTests {
    // MARK: - Internal

    // MARK: - Accessibility Labels on Interactive Elements

    @Test("Interactive elements have accessibility modifiers")
    func interactiveElements_haveAccessibilityModifiers() {
        // Find all screen files (presentation/screen/)
        let screenFiles = SourceFileScanner.swiftFiles(in: sourceRoot)
            .filter { $0.path.contains("/screen/") || $0.path.contains("/Screen/") || $0.path.contains("/Screen.swift")
            }

        var missingAccessibility: [SourceFinding] = []

        for file in screenFiles {
            let findings = findImagesWithoutAccessibility(in: file)
            missingAccessibility.append(contentsOf: findings)
        }

        // Warning-level: log issues but don't fail the build
        if !missingAccessibility.isEmpty {
            // swiftlint:disable:next no_print_statements
            print("⚠️ Found \(missingAccessibility.count) image(s) without accessibility modifiers")
        }
    }

    // MARK: - Touch Target Size

    @Test("No explicit small frame sizes on interactive elements")
    func interactiveElements_meetMinimumTouchTarget() {
        let screenFiles = SourceFileScanner.swiftFiles(in: sourceRoot)
            .filter { $0.path.contains("/screen/") || $0.path.contains("/Screen/") || $0.path.contains("/Component/") }

        var smallTargets: [SourceFinding] = []

        for file in screenFiles {
            let findings = findSmallTouchTargets(in: file)
            smallTargets.append(contentsOf: findings)
        }

        // Warning-level: log issues but don't fail the build
        if !smallTargets.isEmpty {
            // swiftlint:disable:next no_print_statements
            print("⚠️ Found \(smallTargets.count) potentially small touch target(s)")
        }
    }

    // MARK: - Color Contrast (Design System Check)

    @Test("Design system components use semantic colors (not hardcoded hex)")
    func designSystem_usesSemanticColors() {
        let featureDir = sourceRoot + "/Feature"
        guard FileManager.default.fileExists(atPath: featureDir) else {
            return
        }

        let hardcodedColors = SourceFileScanner.containsPattern(
            "Color(#colorLiteral",
            in: featureDir,
            excludingPaths: ["Tests", "Preview"],
        ) + SourceFileScanner.containsPattern(
            "Color(red:",
            in: featureDir,
            excludingPaths: ["Tests", "Preview"],
        ) + SourceFileScanner.containsPattern(
            "Color(hex:",
            in: featureDir,
            excludingPaths: ["Tests", "Preview"],
        )

        // Warning-level: log issues but don't fail the build
        if !hardcodedColors.isEmpty {
            // swiftlint:disable:next no_print_statements
            print("⚠️ Found \(hardcodedColors.count) hardcoded color(s) in feature code — use XGColors semantic tokens")
        }
    }

    // MARK: - Private

    private let sourceRoot = SecurityTestHelpers.sourceRoot

    private let minimumTouchTarget: Double = 44

    // MARK: - Helpers

    private func extractNumber(from string: String) -> Double? {
        let trimmed = string.trimmingCharacters(in: .whitespaces)
        var numStr = ""
        for char in trimmed {
            if char.isNumber || char == "." {
                numStr.append(char)
            } else {
                break
            }
        }
        return Double(numStr)
    }

    private func findImagesWithoutAccessibility(in file: URL) -> [SourceFinding] {
        guard let content = try? String(contentsOf: file, encoding: .utf8) else {
            return []
        }
        let lines = content.components(separatedBy: .newlines)
        var findings: [SourceFinding] = []
        let lookAheadCount = 5

        for (index, line) in lines.enumerated() {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard trimmed.contains("Image("), !trimmed.hasPrefix("//") else {
                continue
            }
            let end = min(index + 1 + lookAheadCount, lines.count)
            let lookAhead = lines[min(index + 1, lines.count - 1) ..< end].joined(separator: " ")
            let hasA11y = lookAhead.contains("accessibilityLabel") ||
                lookAhead.contains("accessibilityHidden") ||
                lookAhead.contains(".decorative")
            if !hasA11y {
                findings.append(SourceFinding(file: file.path, line: index + 1, content: trimmed))
            }
        }
        return findings
    }

    private func isBelowMinimum(_ text: String, dimension: String) -> Bool {
        guard let range = text.range(of: dimension) else {
            return false
        }
        guard let value = extractNumber(from: String(text[range.upperBound...])) else {
            return false
        }
        return value < minimumTouchTarget
    }

    private func findSmallTouchTargets(in file: URL) -> [SourceFinding] {
        guard let content = try? String(contentsOf: file, encoding: .utf8) else {
            return []
        }
        let lines = content.components(separatedBy: .newlines)
        var findings: [SourceFinding] = []
        let contextRange = 3

        for (index, line) in lines.enumerated() {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard trimmed.contains(".frame(") else {
                continue
            }
            let context = lines[max(0, index - contextRange) ..< min(index + contextRange, lines.count)]
                .joined(separator: " ")
            let isInteractive = context.contains("Button") || context.contains("onTapGesture") || context
                .contains("tap")
            guard isInteractive else {
                continue
            }

            let finding = SourceFinding(file: file.path, line: index + 1, content: trimmed)
            if isBelowMinimum(trimmed, dimension: "width:") {
                findings.append(finding)
            }
            if isBelowMinimum(trimmed, dimension: "height:") {
                findings.append(finding)
            }
        }
        return findings
    }
}

File: android/app/src/test/java/com/xirigo/ecommerce/accessibility/AccessibilityTests.kt
package com.xirigo.ecommerce.accessibility

import org.junit.Test
import java.io.File
import com.xirigo.ecommerce.infrastructure.SourceFileScanner
import com.xirigo.ecommerce.infrastructure.SourceFinding

/**
 * Static accessibility audit tests that scan source code for common issues.
 * Maps to ANDROID_TEST.md Section 8 — Accessibility Test Layer.
 *
 * Runtime accessibility audits require instrumented tests.
 * These tests catch missing accessibility modifiers at the code level.
 */
class AccessibilityTests {

    private val sourceRoot = File("src/main/java/com/xirigo/ecommerce")
    private val lookAheadCount = 5
    private val contextRange = 3
    private val minimumTouchTargetDp = 44

    @Test
    fun `interactive composables have content descriptions`() {
        val screenFiles = SourceFileScanner.kotlinFiles(sourceRoot)
            .filter { file ->
                val path = file.path.replace(File.separatorChar, '/')
                path.contains("/screen/") || path.contains("/Screen/") || path.endsWith("Screen.kt")
            }

        val missingDescriptions = mutableListOf<SourceFinding>()

        for (file in screenFiles) {
            missingDescriptions.addAll(findImagesWithoutContentDescription(file))
        }

        // Warning-level: log issues but don't fail the build
        if (missingDescriptions.isNotEmpty()) {
            println("WARNING: Found ${missingDescriptions.size} Image(s) without contentDescription")
        }
    }

    @Test
    fun `no explicit small sizes on interactive elements`() {
        val screenFiles = SourceFileScanner.kotlinFiles(sourceRoot)
            .filter { file ->
                val path = file.path.replace(File.separatorChar, '/')
                path.contains("/screen/") || path.contains("/component/") || path.contains("/Component/")
            }

        val smallTargets = mutableListOf<SourceFinding>()

        for (file in screenFiles) {
            smallTargets.addAll(findSmallTouchTargets(file))
        }

        // Warning-level: log issues but don't fail the build
        if (smallTargets.isNotEmpty()) {
            println("WARNING: Found ${smallTargets.size} potentially small touch target(s)")
        }
    }

    @Test
    fun `design system components use theme colors - not hardcoded hex`() {
        val featureDir = File(sourceRoot, "feature")
        if (!featureDir.exists()) return

        val hardcodedColors = SourceFileScanner.containsPattern(
            "Color(0x",
            featureDir,
            excludingPaths = listOf("Test", "Preview"),
        ) + SourceFileScanner.containsPattern(
            "Color(red",
            featureDir,
            excludingPaths = listOf("Test", "Preview"),
        )

        if (hardcodedColors.isNotEmpty()) {
            println(
                "WARNING: Found ${hardcodedColors.size} hardcoded color(s) in feature code — " +
                    "use XGColors tokens instead",
            )
        }
    }

    private fun findImagesWithoutContentDescription(file: File): List<SourceFinding> {
        val lines = file.readLines()
        val findings = mutableListOf<SourceFinding>()

        for ((index, line) in lines.withIndex()) {
            val trimmed = line.trim()
            if (!trimmed.contains("Image(") || trimmed.startsWith("//")) continue

            val end = minOf(index + 1 + lookAheadCount, lines.size)
            val lookAhead = lines.subList(minOf(index + 1, lines.size - 1), end).joinToString(" ")
            val hasA11y = lookAhead.contains("contentDescription") ||
                lookAhead.contains("decorative")
            if (!hasA11y) {
                findings.add(SourceFinding(file = file.path, line = index + 1, content = trimmed))
            }
        }
        return findings
    }

    private fun findSmallTouchTargets(file: File): List<SourceFinding> {
        val lines = file.readLines()
        val findings = mutableListOf<SourceFinding>()

        for ((index, line) in lines.withIndex()) {
            val trimmed = line.trim()
            if (!containsSizeModifier(trimmed)) continue
            if (!isNearInteractiveElement(lines, index)) continue

            val finding = SourceFinding(file = file.path, line = index + 1, content = trimmed)
            if (hasSmallDimension(trimmed)) {
                findings.add(finding)
            }
        }
        return findings
    }

    private fun containsSizeModifier(line: String): Boolean = line.contains(".size(") || line.contains(".width(") ||
        line.contains(".height(") || line.contains("Modifier.size")

    private fun isNearInteractiveElement(lines: List<String>, index: Int): Boolean {
        val start = maxOf(0, index - contextRange)
        val end = minOf(index + contextRange, lines.size)
        val context = lines.subList(start, end).joinToString(" ")
        return context.contains("Button") || context.contains("clickable") ||
            context.contains("onClick")
    }

    private fun hasSmallDimension(text: String): Boolean =
        isBelowMinimum(text, "size(") || isBelowMinimum(text, "width(") ||
            isBelowMinimum(text, "height(")

    private fun isBelowMinimum(text: String, dimension: String): Boolean {
        val idx = text.indexOf(dimension)
        if (idx < 0) return false
        val after = text.substring(idx + dimension.length)
        val numStr = after.takeWhile { it.isDigit() || it == '.' }
        val value = numStr.toDoubleOrNull() ?: return false
        return value < minimumTouchTargetDp
    }
}
